luarules/gadgets/unit_custom_weapons_behaviours.lua

```diff
@@ -265,51 +264,91 @@ end
 -- Based on retarget
 -- Uses no weapon customParams.
 
-resultCaches.guidance = {} -- ownerID = <isFiring, guidanceType, userTarget, guidanceTarget>
-
-specialEffectFunction.guidance = function(projectileID)
-	if spGetProjectileTimeToLive(projectileID) > 0 then
-		local ownerID = spGetProjectileOwnerID(projectileID)
+-- Guidance weapon must be the primary and have burst/reload > 1 frame.
+-- This is hackish but works well to prevent spammy retargeting anyway.
+weaponCustomParamKeys.guidance = {
+	guidance_miss_radius = toPositiveNumber,
+}
 
-		if spGetUnitIsDead(ownerID) ~= false then
-			return true
+local function guidanceLost(projectileID, radius, targetID)
+	local tx, ty, tz
+
+	if radius > 0 then
+		local _, _, _, ux, uy, uz = spGetUnitPosition(targetID, false, true)
+		local elevation = max(spGetGroundHeight(ux, uz), 0)
+		local dx, dy, dz, slope = spGetGroundNormal(ux, uz, true)
+		local swerveRadius = radius * (0.25 + 0.75 * math_random())
+		local swerveAngle = math_tau * math_random()
+		local cosAngle = math_cos(swerveAngle)
+		local sinAngle = math_sin(swerveAngle)
+
+		if elevation <= 0 or slope <= 0.1 then
+			-- Scatter within a ring in the XZ plane.
+			tx = ux + swerveRadius * cosAngle
+			ty = uy
+			tz = uz + swerveRadius * sinAngle
+		else
+			-- Scatter within a ring rotated to align with terrain.
+			local ax, ay, az = 0, 1, 0
+			if dy >= 0.99 then ax, ay = 1, 0 end
+			local bx = ay * dz - az * dy
+			local by = az * dx - ax * dz
+			local bz = ax * dy - ay * dx
+			local cx = dy * bz - dz * by
+			local cy = dz * bx - dx * bz
+			local cz = dx * by - dy * bx
```

Comment:
Isn’t the rotation math here supposed to be `cosAngle * bx`? As written you’re doing `cosAngle + bx`, which adds a scalar instead of scaling the basis vector, so the wobble gets a constant offset in the b direction and becomes biased/off-center instead of a proper rotation. This is the only real breaking problem here. I could see this slipping through if we just fired from high enough up in testing, or against not extreme enough terrain?